*** configure.in.org	Sun Jun 24 00:28:54 2007
--- configure.in	Sun Jun 24 00:29:19 2007
***************
*** 11,22 ****
  AC_PROG_CC
  AC_PROG_RANLIB
  
- # Checks for libraries.
- PKG_CHECK_MODULES([GSSAPI], [libgssapi >= 0.9], [],
-   [AC_MSG_ERROR([Unable to locate information required to use libgssapi.
-   If you have pkgconfig installed, you might try setting environment
-   variable PKG_CONFIG_PATH to /usr/local/lib/pkgconfig])])
- 
  # Checks for header files.
  AC_HEADER_STDC
  AC_CHECK_HEADERS([stddef.h stdlib.h string.h unistd.h])
--- 11,16 ----
*** librpcsecgss.pc.in.org	Sun Jun 24 00:44:35 2007
--- librpcsecgss.pc.in	Sun Jun 24 00:44:49 2007
***************
*** 5,11 ****
  
  Name: librpcsecgss
  Description: Library that implements rpcsec_gss interface.
- Requires: libgssapi
  Version: @PACKAGE_VERSION@
  Libs: -L@libdir@ -lrpcsecgss
  Cflags: -I@includedir@/rpcsecgss
--- 5,10 ----
