diff -Naru ekg2-99999999.orig/plugins/sim/main.c ekg2-99999999/plugins/sim/main.c
--- ekg2-99999999.orig/plugins/sim/main.c	2010-06-13 16:01:12.859850517 +0200
+++ ekg2-99999999/plugins/sim/main.c	2010-06-13 16:50:52.268834362 +0200
@@ -36,7 +36,7 @@
 	char **message = va_arg(ap, char**);
 	int *encrypted = va_arg(ap, int*);
 
-	char *result;
+	char *result, *recoded;
 
 	if (!session || !message || !encrypted)
 		return 0;
@@ -53,7 +53,9 @@
 		return 0;
 	}
 
-	result = sim_message_encrypt((unsigned char*) *message, *recipient);
+	recoded = ekg_locale_to_cp_dup((unsigned char*) *message);
+	result = sim_message_encrypt(recoded, *recipient);
+	xfree(recoded);
 
 	if (!result) {
 		debug("[sim] encryption failed: %s\n", sim_strerror(sim_errno));
