*** firegl_public.c~	Sun Apr 23 10:14:17 2006
--- firegl_public.c	Sun Apr 23 10:21:13 2006
***************
*** 2498,2504 ****
  
  static irqreturn_t ke_irq_handler_wrap(int irq, void *arg1, struct pt_regs *regs)
  {
!     irq_handler_func(irq, arg1, regs);
      return IRQ_HANDLED;
  }
  
--- 2498,2505 ----
  
  static irqreturn_t ke_irq_handler_wrap(int irq, void *arg1, struct pt_regs *regs)
  {
!     if (irq_handler_func)
! 	irq_handler_func(irq, arg1, regs);
      return IRQ_HANDLED;
  }
  
