*** backend/Makefile.in    2006-09-05 23:43:42.512796000 +0200
--- backend/Makefile.in.new    2006-06-09 18:33:05.237058500 +0200
***************
*** 74,80 ****
  LIBS = $(addprefix libsane-,$(addsuffix .la,$(ALL_BACKENDS)))
  CONFIGS = $(addsuffix .conf,$(ALL_BACKENDS)) saned.conf

! EXTRA = sane_strstatus.lo ../sanei/sanei_init_debug.lo ../sanei/sanei_config.lo

  # With libtool-1.0, we have to mention each library object explicitly... ;-(
  ifneq (@LIBOBJS@ @ALLOCA@,)
--- 74,80 ----
  LIBS = $(addprefix libsane-,$(addsuffix .la,$(ALL_BACKENDS)))
  CONFIGS = $(addsuffix .conf,$(ALL_BACKENDS)) saned.conf

! EXTRA = sane_strstatus.lo ../sanei/sanei_init_debug.lo ../sanei/sanei_config.lo ../sanei/sanei_usb.lo

  # With libtool-1.0, we have to mention each library object explicitly... ;-(
  ifneq (@LIBOBJS@ @ALLOCA@,)
