*** src/squidclam.c	Fri Aug 11 01:06:33 2006
--- src/squidclam.c.new	Fri Aug 11 01:17:45 2006
***************
*** 130,135 ****
--- 130,136 ----
          char src[255];   /* don't forget to also change scanf if you tweak this */
          char ident[255]; /* don't forget to also change scanf if you tweak this */
          char method[31]; /* don't forget to also change scanf if you tweak this */
+         char urlgroup[255]; /* don't forget to also change scanf if you tweak this */
      } rq;
  
      CURL *eh;
***************
*** 252,261 ****
  
  	
      /* loop to parse the given URLs */
!     while (scanf("%10000s %255s %255s %31s", rq.url, rq.src, rq.ident, rq.method) != EOF) {
          /* show which url we're handling */
  #ifdef DEBUG
!         syslog(LOG_INFO,"handle url (%.10000s) src=(%.255s) ident=(%.255s) method=(%.31s)\n", rq.url, rq.src, rq.ident, rq.method);
  #endif
          /* quick "fix" for the POST Problem */
          if (rq.method[0] != 'G') {
--- 253,262 ----
  
  	
      /* loop to parse the given URLs */
!     while (scanf("%10000s %255s %255s %31s %255s", rq.url, rq.src, rq.ident, rq.method, rq.urlgroup) != EOF) {
          /* show which url we're handling */
  #ifdef DEBUG
!         syslog(LOG_INFO,"handle url (%.10000s) src=(%.255s) ident=(%.255s) method=(%.31s) urlgroup=(%.255s)\n", rq.url, rq.src, rq.ident, rq.method, rq.urlgroup);
  #endif
          /* quick "fix" for the POST Problem */
          if (rq.method[0] != 'G') {
***************
*** 332,339 ****
  		}
  		else {
  			    fprintf(stdout,"%.1000s?url=%.10000s&virus=%.255s %.255s %.255s %.31s\n", cfg.url, rq.url, virname, rq.src, rq.ident, rq.method);
! 				syslog(LOG_WARNING, "INFECTED url=%.10000s virus=%.255s src=%.255s ident=%.255s method=%.255s",
! 						rq.url, virname, rq.src, rq.ident, rq.method);
  #ifdef DEBUG
  	        	syslog(LOG_WARNING, "Size of file: %i\n", mem.size);
  #endif
--- 332,339 ----
  		}
  		else {
  			    fprintf(stdout,"%.1000s?url=%.10000s&virus=%.255s %.255s %.255s %.31s\n", cfg.url, rq.url, virname, rq.src, rq.ident, rq.method);
! 				syslog(LOG_WARNING, "INFECTED url=%.10000s virus=%.255s src=%.255s ident=%.255s method=%.255s urlgroup=%.255s",
! 						rq.url, virname, rq.src, rq.ident, rq.method, rq.urlgroup);
  #ifdef DEBUG
  	        	syslog(LOG_WARNING, "Size of file: %i\n", mem.size);
  #endif
