diff -Narc squidclamav/config.c squidclamav.new/config.c
*** squidclamav/config.c	Mon Aug 28 09:25:11 2006
--- squidclamav.new/config.c	Thu Feb  1 11:32:26 2007
***************
*** 49,54 ****
--- 49,55 ----
  extern char *redirect_url;	/* from main.c */
  extern char *squidguard;	/* from main.c */
  extern int debug;		/* from main.c */
+ extern int size_limit;		/* from main.c */
  extern int force;		/* from main.c */
  extern int statit;		/* from main.c */
  extern char *log_file;		/* from main.c */
***************
*** 208,213 ****
--- 209,219 ----
  		return;
  	}
  
+ 	if(strcmp(type, "size") == 0) {
+ 		size_limit = atoi(first);
+ 		return;
+ 	}
+ 
  	/* Path to log file */
  	if(strcmp(type, "logfile") == 0) {
  		log_file = (char *)malloc(sizeof(char) * (strlen(first) +1));
diff -Narc squidclamav/main.c squidclamav.new/main.c
*** squidclamav/main.c	Thu Feb  1 11:39:56 2007
--- squidclamav.new/main.c	Thu Feb  1 11:28:29 2007
***************
*** 83,88 ****
--- 83,89 ----
  int interactive = 0;
  int check_content_type = 0;
  int has_regex = 0;
+ int size_limit = 0;
  int debug = 0;
  int force = 0;
  int statit = 0;
***************
*** 411,417 ****
  	}
  	if (debug != 0)
  	  logit(log_file, "DEBUG File size is %'.2f\n", usize);
! 
  	/* get the CONTENT_TYPE out of the header */
  	if (check_content_type == 1)
  	{
--- 412,426 ----
  	}
  	if (debug != 0)
  	  logit(log_file, "DEBUG File size is %'.2f\n", usize);
! 	
! 	/* Check content-length limit */
! 	if (usize > size_limit) {
! 	  if (debug != 0)
! 	    logit(log_file, "DEBUG File is to big for scanning\n", usize);
! 	  fprintf (stdout, "\n");
! 	  continue;
! 	}
! 	
  	/* get the CONTENT_TYPE out of the header */
  	if (check_content_type == 1)
  	{
