--- src/auth.c.orig	2007-11-19 19:10:01.000000000 +0100
+++ src/auth.c	2007-11-19 19:10:30.000000000 +0100
@@ -362,16 +362,6 @@
 		}
 	}
 
-	/* If we got this far, check the target user's .k5login file. */
-	if ((retval == PAM_SUCCESS) && options->user_check) {
-		if (krb5_kuserok(ctx, userinfo->principal_name, user) == 0) {
-			notice("account checks fail for '%s': user disallowed "
-			       "by .k5login file for '%s'",
-			       userinfo->unparsed_name, user);
-			retval = PAM_PERM_DENIED;
-		}
-	}
-
 	/* Log the authentication status, optionally saving the credentials in
 	 * a piece of shared memory. */
 	if (retval == PAM_SUCCESS) {
