*** pam_krb5afs.c.org	Sat Jun 23 12:12:13 2007
--- pam_krb5afs.c	Sat Jun 23 12:11:31 2007
***************
*** 1309,1315 ****
  			   krb5_max_skdc_timeout, &krb5_max_skdc_timeout);
  	DEBUG("setting maximum timeout to %d", krb5_max_skdc_timeout);
  #endif
! #ifdef HEIMDAL
  	appdefault_integer(context, "heimdal_kdc_timeout", argc, argv,
  		(int) context->kdc_timeout, (int *) &context->kdc_timeout);
  	DEBUG("setting heimdal kdc timeout to %d", context->kdc_timeout);
--- 1309,1315 ----
  			   krb5_max_skdc_timeout, &krb5_max_skdc_timeout);
  	DEBUG("setting maximum timeout to %d", krb5_max_skdc_timeout);
  #endif
! #if 0
  	appdefault_integer(context, "heimdal_kdc_timeout", argc, argv,
  		(int) context->kdc_timeout, (int *) &context->kdc_timeout);
  	DEBUG("setting heimdal kdc timeout to %d", context->kdc_timeout);
