Backported fixes from php-5.3 to php-5.2.6

diff -Naru php-5.2.6/ext/interbase/ibase_blobs.c php5.3/ext/interbase/ibase_blobs.c
--- php-5.2.6/ext/interbase/ibase_blobs.c	2007-12-31 08:20:07.000000000 +0100
+++ php5.3/ext/interbase/ibase_blobs.c	2007-12-31 08:17:09.000000000 +0100
@@ -16,7 +16,7 @@
    +----------------------------------------------------------------------+
  */
 
-/* $Id: ibase_blobs.c,v 1.9.2.1.2.4 2007/12/31 07:20:07 sebastian Exp $ */
+/* $Id: ibase_blobs.c,v 1.9.2.1.2.3.2.1 2007/12/31 07:17:09 sebastian Exp $ */
 
 #ifdef HAVE_CONFIG_H
 #include "config.h"
diff -Naru php-5.2.6/ext/interbase/ibase_events.c php5.3/ext/interbase/ibase_events.c
--- php-5.2.6/ext/interbase/ibase_events.c	2007-12-31 08:20:07.000000000 +0100
+++ php5.3/ext/interbase/ibase_events.c	2007-12-31 08:17:09.000000000 +0100
@@ -16,7 +16,7 @@
    +----------------------------------------------------------------------+
  */
 
-/* $Id: ibase_events.c,v 1.8.2.1.2.2 2007/12/31 07:20:07 sebastian Exp $ */
+/* $Id: ibase_events.c,v 1.8.2.1.2.1.2.1 2007/12/31 07:17:09 sebastian Exp $ */
 
 #ifdef HAVE_CONFIG_H
 #include "config.h"
diff -Naru php-5.2.6/ext/interbase/ibase_query.c php5.3/ext/interbase/ibase_query.c
--- php-5.2.6/ext/interbase/ibase_query.c	2007-12-31 08:20:07.000000000 +0100
+++ php5.3/ext/interbase/ibase_query.c	2007-12-31 08:17:09.000000000 +0100
@@ -16,7 +16,7 @@
    +----------------------------------------------------------------------+
  */
 
-/* $Id: ibase_query.c,v 1.23.2.1.2.11 2007/12/31 07:20:07 sebastian Exp $ */
+/* $Id: ibase_query.c,v 1.23.2.1.2.10.2.5 2007/12/31 07:17:09 sebastian Exp $ */
 
 #ifdef HAVE_CONFIG_H
 #include "config.h"
@@ -49,6 +49,7 @@
 typedef struct {
 	ibase_db_link *link;
 	ibase_trans *trans;
+	struct _ib_query *query;
 	isc_stmt_handle stmt;
 	unsigned short type;
 	unsigned char has_more_rows, statement_type;
@@ -56,9 +57,10 @@
 	ibase_array out_array[1]; /* last member */
 } ibase_result;
 
-typedef struct {
+typedef struct _ib_query {
 	ibase_db_link *link;
 	ibase_trans *trans;
+	ibase_result *result;
 	int result_res_id;
 	isc_stmt_handle stmt;
 	XSQLDA *in_sqlda, *out_sqlda;
@@ -116,6 +118,24 @@
 }
 /* }}} */
 
+static void _php_ibase_free_stmt_handle(ibase_db_link *link, isc_stmt_handle stmt TSRMLS_DC) /* {{{ */
+{
+	static char info[] = { isc_info_base_level, isc_info_end };
+
+	if (stmt) {
+		char res_buf[8];
+		IBDEBUG("Dropping statement handle (free_stmt_handle)...");
+		/* Only free statement if db-connection is still open */
+		if (SUCCESS == isc_database_info(IB_STATUS, &link->handle, 
+							sizeof(info), info, sizeof(res_buf), res_buf)) {
+			if (isc_dsql_free_statement(IB_STATUS, &stmt, DSQL_drop)) {
+				_php_ibase_error(TSRMLS_C);
+			}
+		}
+	}
+}
+/* }}} */
+
 static void _php_ibase_free_result(zend_rsrc_list_entry *rsrc TSRMLS_DC) /* {{{ */
 {
 	ibase_result *ib_result = (ibase_result *) rsrc->ptr;
@@ -123,9 +143,11 @@
 	IBDEBUG("Freeing result by dtor...");
 	if (ib_result) {
 		_php_ibase_free_xsqlda(ib_result->out_sqlda);
-		if (ib_result->stmt && ib_result->type != EXECUTE_RESULT) {
-			IBDEBUG("Dropping statement handle (free_result dtor)...");
-			isc_dsql_free_statement(IB_STATUS, &ib_result->stmt, DSQL_drop);
+		if (ib_result->query != NULL) {
+			IBDEBUG("query still valid; don't drop statement handle");
+			ib_result->query->result = NULL;	/* Indicate to query, that result is released */
+		} else {
+			_php_ibase_free_stmt_handle(ib_result->link, ib_result->stmt TSRMLS_CC);
 		}
 		efree(ib_result);
 	}
@@ -142,11 +164,11 @@
 	if (ib_query->out_sqlda) {
 		efree(ib_query->out_sqlda);
 	}
-	if (ib_query->stmt) {
-		IBDEBUG("Dropping statement handle (free_query)...");
-		if (isc_dsql_free_statement(IB_STATUS, &ib_query->stmt, DSQL_drop)) {
-			_php_ibase_error(TSRMLS_C);
-		}
+	if (ib_query->result != NULL) {
+		IBDEBUG("result still valid; don't drop statement handle");
+		ib_query->result->query = NULL;	/* Indicate to result, that query is released */
+	} else {
+		_php_ibase_free_stmt_handle(ib_query->link, ib_query->stmt TSRMLS_CC);
 	}
 	if (ib_query->in_array) {
 		efree(ib_query->in_array);
@@ -166,11 +188,7 @@
 
 	if (ib_query != NULL) {
 		IBDEBUG("Preparing to free query by dtor...");
-
 		_php_ibase_free_query(ib_query TSRMLS_CC);
-		if (ib_query->statement_type != isc_info_sql_stmt_exec_procedure) {
-			zend_list_delete(ib_query->result_res_id);
-		}
 		efree(ib_query);
 	}
 }
@@ -302,9 +320,16 @@
 	static char info_type[] = {isc_info_sql_stmt_type};
 	char result[8];
 
+	/* Return FAILURE, if querystring is empty */
+	if (*query == '\0') {
+		php_error_docref(NULL TSRMLS_CC, E_WARNING, "Querystring empty.");
+		return FAILURE;
+	}
+
 	ib_query->link = link;
 	ib_query->trans = trans;
 	ib_query->result_res_id = 0;
+	ib_query->result = NULL;
 	ib_query->stmt = NULL;
 	ib_query->in_array = NULL;
 	ib_query->out_array = NULL;
@@ -929,7 +954,10 @@
 		res = emalloc(sizeof(ibase_result)+sizeof(ibase_array)*max(0,ib_query->out_array_cnt-1));
 		res->link = ib_query->link;
 		res->trans = ib_query->trans;
-		res->stmt = ib_query->stmt; 
+		res->stmt = ib_query->stmt;
+		/* ib_result and ib_query point at each other to handle release of statement handle properly */
+		res->query = ib_query;
+		ib_query->result = res;
 		res->statement_type = ib_query->statement_type;
 		res->has_more_rows = 1;
 
@@ -1288,9 +1316,23 @@
 static int _php_ibase_var_zval(zval *val, void *data, int type, int len, /* {{{ */
 	int scale, int flag TSRMLS_DC)
 {
-	static ISC_INT64 const scales[] = { 1, 10, 100, 1000, 10000, 100000, 1000000, 100000000, 1000000000, 
-		1000000000, LL_LIT(10000000000),LL_LIT(100000000000),LL_LIT(10000000000000),LL_LIT(100000000000000),
-		LL_LIT(1000000000000000),LL_LIT(1000000000000000),LL_LIT(1000000000000000000) };
+	static ISC_INT64 const scales[] = { 1, 10, 100, 1000, 
+		10000, 
+		100000, 
+		1000000, 
+		10000000,
+		100000000, 
+		1000000000, 
+		LL_LIT(10000000000), 
+		LL_LIT(100000000000),
+		LL_LIT(1000000000000), 
+		LL_LIT(10000000000000), 
+		LL_LIT(100000000000000),
+		LL_LIT(1000000000000000),
+		LL_LIT(10000000000000000), 
+		LL_LIT(100000000000000000), 
+		LL_LIT(1000000000000000000)
+	};
 
 	switch (type & ~1) {
 		unsigned short l;
diff -Naru php-5.2.6/ext/interbase/ibase_service.c php5.3/ext/interbase/ibase_service.c
--- php-5.2.6/ext/interbase/ibase_service.c	2007-12-31 08:20:07.000000000 +0100
+++ php5.3/ext/interbase/ibase_service.c	2007-12-31 08:17:09.000000000 +0100
@@ -16,7 +16,7 @@
    +----------------------------------------------------------------------+
  */
 
-/* $Id: ibase_service.c,v 1.11.2.2.2.6 2007/12/31 07:20:07 sebastian Exp $ */
+/* $Id: ibase_service.c,v 1.11.2.2.2.5.2.1 2007/12/31 07:17:09 sebastian Exp $ */
 
 #ifdef HAVE_CONFIG_H
 #include "config.h"
diff -Naru php-5.2.6/ext/interbase/interbase.c php5.3/ext/interbase/interbase.c
--- php-5.2.6/ext/interbase/interbase.c	2007-12-31 08:20:07.000000000 +0100
+++ php5.3/ext/interbase/interbase.c	2008-03-10 23:12:34.000000000 +0100
@@ -18,7 +18,7 @@
    +----------------------------------------------------------------------+
  */
 
-/* $Id: interbase.c,v 1.225.2.4.2.8 2007/12/31 07:20:07 sebastian Exp $ */
+/* $Id: interbase.c,v 1.225.2.4.2.7.2.3 2008/03/10 22:12:34 felipe Exp $ */
 
 #ifdef HAVE_CONFIG_H
 #include "config.h"
@@ -47,7 +47,7 @@
 static PHP_GINIT_FUNCTION(ibase);
 
 /* {{{ extension definition structures */
-zend_function_entry ibase_functions[] = {
+const zend_function_entry ibase_functions[] = {
 	PHP_FE(ibase_connect, NULL)
 	PHP_FE(ibase_pconnect, NULL)
 	PHP_FE(ibase_close, NULL)
diff -Naru php-5.2.6/ext/interbase/php_ibase_includes.h php5.3/ext/interbase/php_ibase_includes.h
--- php-5.2.6/ext/interbase/php_ibase_includes.h	2008-01-23 02:22:57.000000000 +0100
+++ php5.3/ext/interbase/php_ibase_includes.h	2008-01-23 02:22:24.000000000 +0100
@@ -18,7 +18,7 @@
    +----------------------------------------------------------------------+
  */
 
-/* $Id: php_ibase_includes.h,v 1.16.2.1.2.3 2008/01/23 01:22:57 iliaa Exp $ */
+/* $Id: php_ibase_includes.h,v 1.16.2.1.2.1.2.2 2008/01/23 01:22:24 iliaa Exp $ */
 
 #ifndef PHP_IBASE_INCLUDES_H
 #define PHP_IBASE_INCLUDES_H
diff -Naru php-5.2.6/ext/interbase/php_ibase_udf.c php5.3/ext/interbase/php_ibase_udf.c
--- php-5.2.6/ext/interbase/php_ibase_udf.c	2007-12-31 08:20:07.000000000 +0100
+++ php5.3/ext/interbase/php_ibase_udf.c	2007-12-31 08:17:09.000000000 +0100
@@ -16,7 +16,7 @@
    +----------------------------------------------------------------------+
  */
 
-/* $Id: php_ibase_udf.c,v 1.9.2.1.2.3 2007/12/31 07:20:07 sebastian Exp $ */
+/* $Id: php_ibase_udf.c,v 1.9.2.1.2.2.2.1 2007/12/31 07:17:09 sebastian Exp $ */
 
 /**
 * This UDF library adds the ability to call PHP functions from SQL
diff -Naru php-5.2.6/ext/interbase/php_interbase.h php5.3/ext/interbase/php_interbase.h
--- php-5.2.6/ext/interbase/php_interbase.h	2007-12-31 08:20:07.000000000 +0100
+++ php5.3/ext/interbase/php_interbase.h	2007-12-31 08:17:09.000000000 +0100
@@ -18,7 +18,7 @@
    +----------------------------------------------------------------------+
  */
 
-/* $Id: php_interbase.h,v 1.71.2.1.2.2 2007/12/31 07:20:07 sebastian Exp $ */
+/* $Id: php_interbase.h,v 1.71.2.1.2.1.2.1 2007/12/31 07:17:09 sebastian Exp $ */
 
 #ifndef PHP_INTERBASE_H
 #define PHP_INTERBASE_H
