*** agent/mibgroup/ip-mib/ipAddressTable/ipAddressTable_data_access.c	Wed Jan 25 17:25:56 2006
--- agent/mibgroup/ip-mib/ipAddressTable/ipAddressTable_data_access.c.new	Mon Jan 22 18:43:23 2007
***************
*** 333,340 ****
       * the container.
       */
      ipaddress_container =
!         netsnmp_access_ipaddress_container_load(NULL,
!                                                 NETSNMP_ACCESS_IPADDRESS_LOAD_ADDL_IDX_BY_ADDR);
      /*
       * we just got a fresh copy of interface data. compare it to
       * what we've already got, and make any adjustments, saving
--- 333,339 ----
       * the container.
       */
      ipaddress_container =
!         netsnmp_access_ipaddress_container_load(NULL, 0);
      /*
       * we just got a fresh copy of interface data. compare it to
       * what we've already got, and make any adjustments, saving
