--- Makefile.config.in.orig	2009-08-10 00:16:48.000000000 +0200
+++ Makefile.config.in	2009-08-10 00:17:13.000000000 +0200
@@ -22,6 +22,7 @@
 ENABLE_CPLUSPLUS       = @ENABLE_CPLUSPLUS@
 ENABLE_CGI_SERVER      = @ENABLE_CGI_SERVER@
 ENABLE_LIBXML2_BACKEND = @ENABLE_LIBXML2_BACKEND@
+ENABLE_LIBXML2         = @ENABLE_LIBXML2_BACKEND@
 ENABLE_EFENCE          = @ENABLE_EFENCE@
 
 MUST_BUILD_WININET_CLIENT = @MUST_BUILD_WININET_CLIENT@
--- Makefile.config.in.orig	2009-08-10 00:19:50.000000000 +0200
+++ Makefile.config.in	2009-08-10 00:20:58.000000000 +0200
@@ -61,3 +61,6 @@
 LIBINST_DIR = @libdir@
 HEADERINST_DIR = @includedir@
 
+LIBXML2_CFLAGS = @LIBXML2_CFLAGS@
+LIBXML2_LIBS   = @LIBXML2_CFLAGS@
+
--- src/XmlRpcCpp.h.orig	2009-08-10 00:25:58.000000000 +0200
+++ src/XmlRpcCpp.h	2009-08-10 00:26:38.000000000 +0200
@@ -28,6 +28,8 @@
 
 
 #include <string>
+#include <cstring>
+#include <cstdlib>
 
 using namespace std;
 
