*** fs/splice.c.orig	Sun Feb 10 17:14:08 2008
--- fs/splice.c	Sun Feb 10 22:45:42 2008
***************
*** 1448,1456 ****
  			break;
  		}
  
! 		sd.len = 0;
! 		sd.total_len = len;
! 		sd.flags = flags;
  		sd.u.userptr = base;
  		sd.pos = 0;
  
--- 1448,1461 ----
  			break;
  		}
  
! 		if (unlikely(!access_ok(VERIFY_WRITE, base, len))) {
! 			error = -EFAULT;
! 			break;
! 		}
! 
!                 sd.len = 0;
!                 sd.total_len = len;
!                 sd.flags = flags;
  		sd.u.userptr = base;
  		sd.pos = 0;
  
