*** Ocsinventory/Agent/Option/Download.pm.org	2007-05-25 17:40:45.000000000 +0200
--- Ocsinventory/Agent/Option/Download.pm	2007-05-25 17:41:09.000000000 +0200
***************
*** 230,236 ****
  				
  				my($server_name,$server_port,$server_dir);
  				
! 				if($_->{INFO_LOC}=~ /^(\w+):(\d{1,5})(.*)$/){
  					$server_name = $1;
  					$server_port = $2;
  					$server_dir = $3;
--- 230,236 ----
  				
  				my($server_name,$server_port,$server_dir);
  				
! 				if($_->{INFO_LOC}=~ /^(\S+):(\d{1,5})(.*)$/){
  					$server_name = $1;
  					$server_port = $2;
  					$server_dir = $3;
***************
*** 666,672 ****
  				}
  				
  				&log("Storing package to $p->{'PATH'}...");
! 				unlink("build.tar.gz") or die("Cannot remove build file: $!");
  				
  				system("cp -r * ".$p->{'PATH'}) and die();
  			}
--- 666,672 ----
  				}
  				
  				&log("Storing package to $p->{'PATH'}...");
! 				unlink("build.tar.gz");
  				
  				system("cp -r * ".$p->{'PATH'}) and die();
  			}
