*** source/utils/net.c.org	2006-09-12 12:32:55.000000000 +0200
--- source/utils/net.c	2006-09-12 12:32:07.000000000 +0200
***************
*** 72,78 ****
  const char *opt_comment = "";
  const char *opt_container = NULL;
  int opt_flags = -1;
! int opt_timeout = 0;
  const char *opt_target_workgroup = NULL;
  int opt_machine_pass = 0;
  BOOL opt_localgroup = False;
--- 72,78 ----
  const char *opt_comment = "";
  const char *opt_container = NULL;
  int opt_flags = -1;
! int opt_timeout = -1;
  const char *opt_target_workgroup = NULL;
  int opt_machine_pass = 0;
  BOOL opt_localgroup = False;
*** source/utils/net_rpc.c	2006-09-12 13:58:40.000000000 +0200
--- source/utils/net_rpc.c.new	2006-09-12 14:03:11.000000000 +0200
***************
*** 5179,5185 ****
  	if (opt_comment) {
  		msg = opt_comment;
  	}
! 	if (opt_timeout) {
  		timeout = opt_timeout;
  	}
  
--- 5179,5185 ----
  	if (opt_comment) {
  		msg = opt_comment;
  	}
! 	if (opt_timeout>=0) {
  		timeout = opt_timeout;
  	}
  
***************
*** 5188,5194 ****
  				   opt_force);
  
  	if (NT_STATUS_IS_OK(result)) {
! 		d_printf("\nShutdown of remote machine succeeded\n");
  		DEBUG(5,("Shutdown of remote machine succeeded\n"));
  	} else {
  		DEBUG(1,("Shutdown of remote machine failed!\n"));
--- 5188,5194 ----
  				   opt_force);
  
  	if (NT_STATUS_IS_OK(result)) {
! 		d_printf("\nShutdown of remote machine in %ds with message '%s' succeeded\n",timeout,msg);
  		DEBUG(5,("Shutdown of remote machine succeeded\n"));
  	} else {
  		DEBUG(1,("Shutdown of remote machine failed!\n"));
***************
*** 5251,5257 ****
  	if (opt_comment) {
  		msg = opt_comment;
  	}
! 	if (opt_timeout) {
  		timeout = opt_timeout;
  	}
  
--- 5251,5257 ----
  	if (opt_comment) {
  		msg = opt_comment;
  	}
! 	if (opt_timeout>=0) {
  		timeout = opt_timeout;
  	}
  
***************
*** 5259,5265 ****
  	result = rpccli_reg_shutdown(pipe_hnd, mem_ctx, msg, timeout, opt_reboot, opt_force);
  
  	if (W_ERROR_IS_OK(result)) {
! 		d_printf("\nShutdown of remote machine succeeded\n");
  	} else {
  		d_fprintf(stderr, "\nShutdown of remote machine failed\n");
  		if (W_ERROR_EQUAL(result,WERR_MACHINE_LOCKED))
--- 5259,5265 ----
  	result = rpccli_reg_shutdown(pipe_hnd, mem_ctx, msg, timeout, opt_reboot, opt_force);
  
  	if (W_ERROR_IS_OK(result)) {
! 		d_printf("\nShutdown of remote machine in %ds with message '%s' succeeded\n",timeout,msg);
  	} else {
  		d_fprintf(stderr, "\nShutdown of remote machine failed\n");
  		if (W_ERROR_EQUAL(result,WERR_MACHINE_LOCKED))
