--- storage/authreg_ldap.c	2007-04-28 23:56:58.000000000 +0200
+++ storage/authreg_ldap.c	2007-08-05 21:55:18.000000000 +0200
@@ -405,9 +405,9 @@
       data->flags |= AR_LDAP_FLAGS_V3;
 
     if (l>0)
-        snprintf(ldap_entry,sizeof(ldap_entry), "authreg.ldap%d.startls", l );
+        snprintf(ldap_entry,sizeof(ldap_entry), "authreg.ldap%d.starttls", l );
     else
-        snprintf(ldap_entry, sizeof(ldap_entry), "authreg.ldap.startls");
+        snprintf(ldap_entry, sizeof(ldap_entry), "authreg.ldap.starttls");
     if(config_get(ar->c2s->config, ldap_entry) != NULL)
       data->flags |= AR_LDAP_FLAGS_STARTTLS;
