diff -u portage-utils-0.2.1.orig/qlop.c portage-utils-0.2.1/qlop.c
--- portage-utils-0.2.1.orig/qlop.c	2009-08-29 21:45:54.000000000 +0200
+++ portage-utils-0.2.1/qlop.c	2009-12-03 22:54:19.000000000 +0100
@@ -469,7 +469,7 @@
 	 * to round up if it can predict a sudden growth, so optimise below
 	 * for the optimistic case */
 	ret = sysctl(mib, 3, NULL, &size, NULL, 0);
-	ip = malloc(sizeof(struct kinfo_proc) * size);
+	ip = xmalloc(sizeof(struct kinfo_proc) * size);
 	if (ip == NULL) {
 		warnp("Could not allocate %d bytes for process information",
 				sizeof(struct kinfo_proc) * size);
