*** frontends/ovdb_stat.c.org	Mon Oct 23 01:20:21 2006
--- frontends/ovdb_stat.c	Mon Oct 23 01:51:13 2006
***************
*** 586,592 ****
  {
      DB_BTREE_STAT *sp;
  
! #if DB_VERSION_MAJOR >= 4 || (DB_VERSION_MAJOR >= 3 && DB_VERSION_MINOR >= 3)
      if(db->stat(db, &sp, 0))
  	return 1;
  #else
--- 586,592 ----
  {
      DB_BTREE_STAT *sp;
  
! #if DB_VERSION_MAJOR > 4 || (DB_VERSION_MAJOR == 3 && DB_VERSION_MINOR >= 3)
      if(db->stat(db, &sp, 0))
  	return 1;
  #else
***************
*** 646,652 ****
  #else
      DB_HASH_STAT *sp;
  
! #if DB_VERSION_MAJOR == 3 && DB_VERSION_MINOR <= 2
      if(db->stat(db, &sp, NULL, 0))
  	return 1;
  #else
--- 646,652 ----
  #else
      DB_HASH_STAT *sp;
  
! #if DB_VERSION_MAJOR == 3 && DB_VERSION_MINOR <= 2 || DB_VERSION_MAJOR == 4
      if(db->stat(db, &sp, NULL, 0))
  	return 1;
  #else
