diff -ru nfs-utils-1.2.2.orig/aclocal/rpcsec_vers.m4 nfs-utils-1.2.2/aclocal/rpcsec_vers.m4
--- nfs-utils-1.2.2.orig/aclocal/rpcsec_vers.m4	2010-02-18 13:35:00.000000000 +0100
+++ nfs-utils-1.2.2/aclocal/rpcsec_vers.m4	2010-05-31 01:25:34.480602835 +0200
@@ -1,7 +1,7 @@
 dnl Checks librpcsec version
 AC_DEFUN([AC_RPCSEC_VERSION], [
 
-  PKG_CHECK_MODULES([GSSGLUE], [libgssglue >= 0.1])
+  PKG_CHECK_MODULES([GSSGLUE], [libgssglue >= 0.1], , [PKG_CHECK_MODULES([GSSGLUE], [heimdal-gssapi])])
 
   dnl TI-RPC replaces librpcsecgss
   if test "$enable_tirpc" = no; then
diff -ru nfs-utils-1.2.2.orig/configure.ac nfs-utils-1.2.2/configure.ac
--- nfs-utils-1.2.2.orig/configure.ac	2010-02-18 13:35:00.000000000 +0100
+++ nfs-utils-1.2.2/configure.ac	2010-05-31 02:03:04.052604578 +0200
@@ -81,7 +81,7 @@
 	if test "$enable_nfsv41" = yes; then
 		AC_DEFINE(NFS41_SUPPORTED, 1, [Define this if you want NFSv41 support compiled in])
 	else
-		enable_nfsv4=
+		enable_nfsv41=
 	fi
 	AC_SUBST(enable_nfsv41)
 	AM_CONDITIONAL(CONFIG_NFSV41, [test "$enable_nfsv41" = "yes"])
