*** ./utils/gssd/krb5_util.c.org	Sun Jun 24 00:55:18 2007
--- ./utils/gssd/krb5_util.c	Sun Jun 24 00:56:46 2007
***************
*** 294,300 ****
  		return -1;
  	}
  
! 	maj_stat = gss_set_allowable_enctypes(&min_stat, credh, &krb5oid,
  					     num_enctypes, &enctypes);
  	if (maj_stat != GSS_S_COMPLETE) {
  		pgsserr("gss_set_allowable_enctypes",
--- 294,300 ----
  		return -1;
  	}
  
! 	maj_stat = gss_krb5_set_allowable_enctypes(&min_stat, credh,
  					     num_enctypes, &enctypes);
  	if (maj_stat != GSS_S_COMPLETE) {
  		pgsserr("gss_set_allowable_enctypes",
