--- utils/gssd/krb5_util.c.orig	2008-09-01 01:10:15.077637090 +0200
+++ utils/gssd/krb5_util.c	2008-09-01 01:33:47.122513890 +0200
@@ -936,13 +936,12 @@
 
 	while (!found &&
 		(ret = krb5_cc_next_cred(context, ccache, &cur, &creds)) == 0) {
-		if (creds.server->length == 2 &&
-				data_is_equal(creds.server->realm,
-					      principal->realm) &&
-				creds.server->data[0].length == 6 &&
-				memcmp(creds.server->data[0].data,
-						"krbtgt", 6) == 0 &&
-				data_is_equal(creds.server->data[1],
-					      principal->realm) &&
+		if (creds.server->name.name_string.len == 2 &&
+				strcmp(creds.server->realm,
+					      principal->realm) == 0 &&
+				strcmp(creds.server->name.name_string.val[0],
+						"krbtgt") == 0 &&
+				strcmp(creds.server->name.name_string.val[1],
+					      principal->realm) == 0 &&
 				creds.times.endtime > time(NULL))
 			found = 1;
@@ -995,7 +994,7 @@
 	}
 	krb5_free_principal(context, principal);
 err_princ:
-	krb5_cc_set_flags(context, ccache,  KRB5_TC_OPENCLOSE);
+//	krb5_cc_set_flags(context, ccache,  KRB5_TC_OPENCLOSE);
 	krb5_cc_close(context, ccache);
 err_cache:
 	krb5_free_context(context);
