--- ./aclocal/rpcsec_vers.m4.orig	2009-05-18 00:57:10.160719435 +0200
+++ ./aclocal/rpcsec_vers.m4	2009-05-18 00:57:28.452314767 +0200
@@ -1,8 +1,6 @@
 dnl Checks librpcsec version
 AC_DEFUN([AC_RPCSEC_VERSION], [
 
-  PKG_CHECK_MODULES([GSSGLUE], [libgssglue >= 0.1])
-
   dnl TI-RPC replaces librpcsecgss
   if test "$enable_tirpc" = no; then
     PKG_CHECK_MODULES([RPCSECGSS], [librpcsecgss >= 0.16])
