Index: nfs-utils-1.1.0/utils/gssd/context_lucid.c
===================================================================
--- nfs-utils-1.1.0.orig/utils/gssd/context_lucid.c
+++ nfs-utils-1.1.0/utils/gssd/context_lucid.c
@@ -48,8 +48,10 @@
 #include "context.h"
 
 #ifndef OM_uint64
+#ifndef GSSAPI_GSSAPI_H_
 typedef uint64_t OM_uint64;
 #endif
+#endif
 
 static int
 write_lucid_keyblock(char **p, char *end, gss_krb5_lucid_key_t *key)
