diff -ru nfs-utils-1.2.0.orig/utils/gssd/krb5_util.c nfs-utils-1.2.0/utils/gssd/krb5_util.c
--- nfs-utils-1.2.0.orig/utils/gssd/krb5_util.c	2009-09-19 23:38:37.000000000 +0200
+++ nfs-utils-1.2.0/utils/gssd/krb5_util.c	2009-09-20 00:00:53.000000000 +0200
@@ -1261,7 +1261,12 @@
 	if (context != NULL) {
 		origmsg = krb5_get_error_message(context, code);
 		msg = strdup(origmsg);
+#ifdef HAVE_HEIMDAL
+		krb5_free_error_string(context, origmsg);
+#else
 		krb5_free_error_message(context, origmsg);
+#endif
+
 	}
 #endif
 	if (msg != NULL)
