diff -Naur nagios-plugins-1.4.15.orig//plugins/netutils.c nagios-plugins-1.4.15/plugins/netutils.c
--- nagios-plugins-1.4.15.orig//plugins/netutils.c	2012-06-25 22:08:38.000000000 +0200
+++ nagios-plugins-1.4.15/plugins/netutils.c	2012-06-25 22:09:36.000000000 +0200
@@ -203,6 +203,7 @@
 		}
 
 		if (source_ip_str != 0) {
+			memset(&source_ip, 0, sizeof(struct sockaddr_in));
 			source_ip.sin_family = AF_INET; // Sorry, no IPv6
 			inet_aton(source_ip_str, (struct in_addr*) &source_ip.sin_addr);
 		}
